compiler/rustc_builtin_macros/src/deriving/default.rs

```diff
@@ -34,53 +37,180 @@ pub fn expand_deriving_default(
-                    cx.expr_call_ident(trait_span, substr.type_ident, exprs)
-                }
-            }
-            Named(ref fields) => {
-                let default_fields = fields
-                    .iter()
-                    .map(|&(ident, span)| cx.field_imm(span, ident, default_call(span)))
-                    .collect();
-                cx.expr_struct_ident(trait_span, substr.type_ident, default_fields)
+    match summary {
+        Unnamed(ref fields, is_tuple) => {
+            if !is_tuple {
+                cx.expr_ident(trait_span, substr.type_ident)
+            } else {
+                let exprs = fields.iter().map(|sp| default_call(*sp)).collect();
+                cx.expr_call_ident(trait_span, substr.type_ident, exprs)
             }
-        },
-        StaticEnum(..) => {
-            struct_span_err!(
-                &cx.sess.parse_sess.span_diagnostic,
-                trait_span,
-                E0665,
-                "`Default` cannot be derived for enums, only structs"
-            )
+        }
+        Named(ref fields) => {
+            let default_fields = fields
+                .iter()
+                .map(|&(ident, span)| cx.field_imm(span, ident, default_call(span)))
+                .collect();
+            cx.expr_struct_ident(trait_span, substr.type_ident, default_fields)
+        }
+    }
+}
+
+fn default_enum_substructure(
+    cx: &mut ExtCtxt<'_>,
+    trait_span: Span,
+    enum_def: &EnumDef,
+    _fields: &[(Ident, Span, StaticFields)],
+) -> P<Expr> {
+    let failed_derive_expr = DummyResult::raw_expr(trait_span, true);
+
+    let default_variant = match extract_default_variant(cx, enum_def, trait_span) {
+        Ok(value) => value,
+        Err(()) => return failed_derive_expr,
+    };
+
+    // At this point, we know that there is exactly one variant with a `#[default]` attribute. The
+    // attribute hasn't yet been validated.
+
+    if let Err(()) = validate_default_attribute(cx, default_variant) {
+        return failed_derive_expr;
+    }
+
+    // We now know there is exactly one unit variant with exactly one `#[default]` attribute.
+
+    cx.expr_path(cx.path(
+        default_variant.span,
+        vec![Ident::new(kw::SelfUpper, default_variant.span), default_variant.ident],
+    ))
+}
+
+fn extract_default_variant<'a>(
+    cx: &mut ExtCtxt<'_>,
+    enum_def: &'a EnumDef,
+    trait_span: Span,
+) -> Result<&'a rustc_ast::Variant, ()> {
+    let default_variants: Vec<_> = enum_def
+        .variants
+        .iter()
+        .filter(|variant| variant.attrs.iter().any(|attr| attr.has_name(kw::Default)))
+        .collect();
+
+    let variant = match default_variants.as_slice() {
+        [variant] => variant,
+        [] => {
+            cx.struct_span_err(trait_span, "no default variant declared")
+                .help("make a unit variant default by placing `#[default]` above it")
+                .emit();
```

Comment:
I only makes sense to report this error if the total number of variants is 2 or more.

Reply:
If the enum has a single variant then the default variant is pretty obvious, if the enum has zero variants then the derive can generate the usual unreachable code stub with empty `match`.

I suggest amending the RFC to support `derive(Default)` without `#[default]` for enums with <2 variants.

Reply:
Sorry, no time to participate in the RFC right now.

Reply:
I'll drop a comment there for that specific case. Personally I don't see it as something that's a big deal given how uncommon univariant enums are. But the error noted here is in conformance with the RFC as it's currently written.